server/src/main/java/org/opensearch/cluster/coordination/JoinHelper.java

```diff
@@ -407,12 +462,35 @@ public String executor() {
     }
 
     public void sendValidateJoinRequest(DiscoveryNode node, ClusterState state, ActionListener<TransportResponse.Empty> listener) {
-        transportService.sendRequest(
-            node,
-            VALIDATE_JOIN_ACTION_NAME,
-            new ValidateJoinRequest(state),
-            new ActionListenerResponseHandler<>(listener, i -> Empty.INSTANCE, ThreadPool.Names.GENERIC)
-        );
+        if (node.getVersion().before(Version.V_2_8_0)) {
+            transportService.sendRequest(
+                node,
+                VALIDATE_JOIN_ACTION_NAME,
+                new ValidateJoinRequest(state),
+                new ActionListenerResponseHandler<>(listener, i -> Empty.INSTANCE, ThreadPool.Names.GENERIC)
+            );
+        } else {
+            try {
+                BytesReference bytes;
+                if (serializedStates.get(state.getVersion()) == null) {
+                    serializedStates.invalidateAll();
+                }
+                bytes = serializedStates.computeIfAbsent(
+                    state.version(),
+                    key -> ClusterStateUtils.serializeClusterState(state, node, true)
+                );
+                final BytesTransportRequest request = new BytesTransportRequest(bytes, node.getVersion());
+                transportService.sendRequest(
+                    node,
+                    VALIDATE_COMPRESSED_JOIN_ACTION_NAME,
+                    request,
+                    new ActionListenerResponseHandler<>(listener, i -> Empty.INSTANCE, ThreadPool.Names.GENERIC)
+                );
+            } catch (Exception e) {
+                logger.warn(() -> new ParameterizedMessage("error sending cluster state to {}", node), e);
+                listener.onFailure(e);
```

Comment:
This is potentially the same thing as log-and-throw. What does the listener do with this failure? Will we get a duplicate stack trace in the logs?

Reply:
Listener.onFailure is necessary.
This listener is passed via Join request, it's basically the transport joinCallBack listener. So, if anything wrong happens, data node should be notified that its join request has failed. 

The flow : 
* cluster manager creates a transportJoinCallback and sets it as part of join handler
* data-node sends a join request to cluster manager
* cluster manager after receiving the join request, tries to connect to the node
* Once connection is successful, cluster manager send the validateJoinRequest
* If anything fails in any of these flow, that transport join call back .failure method needs to be invoked for proper communication back to data node

In case of any other exception apart from transport call of sendValidateJoinRequest, we need to trigger the joinCallBack listener's failure code.

JoinCallback's flow will log that `failed to validate incoming join request from node [{}]`
and validate join(this flow) is logging `error sending cluster state to {node}` --> this log line helps people understand the cluster manager couldn't send cluster state to data node that's why validation of incoming join request from data node failed. Without this log, people who're not aware of code path won't understand why incoming join request from node was not validated.

Listener we need trigger, but log we can skip. Once we skip log, we'll only get **failed to validate incoming join request from node [{}]** in logs.

Just because it's a different transport call from cluster manager to data node, I felt it would be good to have a log line for the failure. Let me know what do you think.

Reply:
Since the listener logs the stack trace, maybe skip that here? Agreed that "error sending cluster state to {node}" can be helpful context but probably not necessary to print the exception stack trace twice.

Reply:
Sure, will print just the log not the trace 👍🏼